homeassistant/components/xiaomi_miio/light.py

```diff
@@ -1031,39 +1030,33 @@ def turn_on(self, **kwargs):
         else:
             brightness_pct = self._brightness_pct
 
-        brightness_and_color = brightness_and_color_to_int(brightness_pct, rgb)
-        self._gateway.send("set_rgb", [brightness_and_color])
+        self._gateway.light.set_rgb(brightness_pct, rgb)
 
         self.schedule_update_ha_state()
 
     def turn_off(self, **kwargs):
         """Turn the light off."""
-        brightness_and_color = brightness_and_color_to_int(0, self._rgb)
-        self._gateway.send("set_rgb", [brightness_and_color])
-
+        self._gateway.light.set_rgb(0, self._rgb)
         self.schedule_update_ha_state()
 
     async def async_update(self):
         """Fetch state from the device."""
         try:
-            state_int = await self.hass.async_add_executor_job(
-                partial(self._gateway.send, "get_rgb")
+            state_dict = await self.hass.async_add_executor_job(
+                self._gateway.light.rgb_status
             )
         except GatewayException as ex:
-            self._available = False
-            _LOGGER.error(
-                "Got exception while fetching the gateway light state: %s", ex
-            )
+            if self._available:
+                self._available = False
+                _LOGGER.error(
+                    "Got exception while fetching the gateway light state: %s", ex
+                )
             return
 
-        state_int = state_int.pop()
         self._available = True
-        brightness_pct = int_to_brightness(state_int)
+        self._is_on = state_dict["is_on"]
 
-        if brightness_pct > 0:
-            self._is_on = True
-            self._brightness_pct = brightness_pct
-            self._rgb = int_to_rgb(state_int)
+        if self._is_on:
+            self._brightness_pct = state_dict["brightness"]
+            self._rgb = state_dict["rgb"]
```

Comment:
Should these also be set to something (None?) when the device is off? Otherwise, please rebase on the current dev (the version bump has already been merged) and this is good to go from my view.

Reply:
No, if the light is off, the brightness and rgb will be reported as 0 and (0,0,0).
If you turn on the light again, we need to know the last brightness and rgb to turn it on to the same brightness and rgb.
Therefore we schould keep the previous values in memory instead of overwriting them by 0.

I will rebase